Added texture type OES

0v3rt1r3d opened the pull request:
Corresponding topic: https://forums.ogre3d.org/viewtopic.php?p=543151#p543151

paroj commented on OgreMain/include/OgreTexture.h:
```diff
@@ -125,6 +127,11 @@ namespace Ogre {
         virtual ~Texture() {}
+
+
+        virtual uint getTextureId() {
```
this is not needed: you can use `getCustomAttribute("GLID", ..)

0v3rt1r3d replied:
I do not understand, how can I get the GLID parameter in java code. Generated method takes `SWIGTYPE_p_void`, but I need to extract integer type

paroj replied:
yeah.. forgot about that, then you need your custom method

paroj replied:
added a wrappable version in #915

paroj commented on RenderSystems/GLES2/src/OgreGLES2RenderSystem.cpp:
```diff
@@ -839,10 +839,14 @@ namespace Ogre {
             return;
 
-        GLenum target = mTextureTypes[unit];
 
```
the early out should probably only be done here so the texture unit still gets activated

0v3rt1r3d replied:
I tried to return my return, it led to INVALID_ENUM error. Than I added correct parameters according to https://www.khronos.org/registry/OpenGL/extensions/OES/OES_EGL_image_external.txt and google example (latest changes). And I still could no see camera image, although no gl error occurs now

0v3rt1r3d commented:
Have you any ideas about my problem? The process of drawing is going on in java code using shaders, maybe an error is in common work with https://github.com/google-ar/arcore-android-sdk/blob/master/samples/hello_ar_java/app/src/main/java/com/google/ar/core/examples/java/common/rendering/BackgroundRenderer.java. I removed part below and added textureId as parameter to constructor:
```
    // Generate the background texture.
    int[] textures = new int[1];
    GLES20.glGenTextures(1, textures, 0);
    textureId = textures[0];
    int textureTarget = GLES11Ext.GL_TEXTURE_EXTERNAL_OES;
    GLES20.glBindTexture(textureTarget, textureId);
    GLES20.glTexParameteri(textureTarget, GLES20.GL_TEXTURE_WRAP_S, GLES20.GL_CLAMP_TO_EDGE);
    GLES20.glTexParameteri(textureTarget, GLES20.GL_TEXTURE_WRAP_T, GLES20.GL_CLAMP_TO_EDGE);
    GLES20.glTexParameteri(textureTarget, GLES20.GL_TEXTURE_MIN_FILTER, GLES20.GL_NEAREST);
    GLES20.glTexParameteri(textureTarget, GLES20.GL_TEXTURE_MAG_FILTER, GLES20.GL_NEAREST);
```

paroj commented:
this could be a different opengl context from the one ogre is using. I would rather try drawing the texture inside ogre: http://wiki.ogre3d.org/Displaying+2D+Backgrounds

for this you only need to slightly adapt the shaders to the ogre naming conventions: https://ogrecave.github.io/ogre/api/latest/_high-level-_programs.html#Binding-vertex-attributes

0v3rt1r3d commented:
I am sure that it is correct opengl context. Android has its own render thread with a context (which I could not access directly), and another context is initialized by ogre. When I run the GAPID tool, I can see both contexts, list of calls for each context (and shaders, and textures etc). So I see my shader and draw commands in the ogre context, also I see texture buffer with correct camera image.

Maybe the render order is important? In java code I call drawing background and then call `renderOneFrame()`

paroj commented:
can you also provide a screenshot of the rendering? it was just black background, right?

renderOneFrame() by default clears the framebuffer, so whatever you render first gets lost. (which is another reason to let ogre render the rectangle)

0v3rt1r3d commented:
Sure, but as I told before, first several frames (one hundred or more) I see another texture (from fish) https://yadi.sk/i/5K3IF2mOPtA9Pg, but then it replaces by black screen https://yadi.sk/i/YOZ1grlnl1zcLw.
How can prevent ogre to clear buffers? I would like to comment some ogre gl-clear calls to achieve the set of instructions in draw method:

- Clear buffers by myself
- draw camera image by myself
- `renderOneFrame()` without clearing buffers

paroj commented:
> How can prevent ogre to clear buffers? 

https://ogrecave.github.io/ogre/api/latest/class_ogre_1_1_viewport.html#ad9c172c23b0dcccd929f31d4416f9683

it is strange that
1. it displays anything in the background at all
2. it uses the wrong texture content.

can you verify whether the id you get from getTextureId is the one you see in GAPID (i.e. 21 and not 22)?

0v3rt1r3d commented:
Verified, it textureId is 21. Tried to call `setClearEveryFrame()` and call `glClear(GL_COLOR_BUFFER_BIT|GL_DEPTH_BUFFER_BIT|GL_STENCIL_BUFFER_BIT)` - nothing changes. (But if I do not clear buffers at all - I stop to see fish).

Maybe I have some errors in creating texture code? (looks strange, because it is kotlin)

```
val material = MaterialManager.getSingleton().create("PlaneMat", ResourceGroupManager.getDEFAULT_RESOURCE_GROUP_NAME())

val textureSmth = material.techniques[0].passes[0].createTextureUnitState()
material.techniques[0].passes[0].apply {
    depthCheckEnabled = false
    depthWriteEnabled = false
    lightingEnabled = false
}

val texture = TextureManager.getSingleton().createManual("SomeTexture", "General", TextureType.TEX_TYPE_EXTERNAL_OES, 400, 400, 0, PixelFormat.PF_R8G8B8)

textureId = texture.textureId // global variable to use in draw-background method, it is wrapped getter to getTextureId() from kotlin

Timber.d("TEXTUREID: got $textureId") // log, which writes 21 (correct texture name)

textureSmth.setTexture(texture)

val rect = Rectangle2D(true)
rect.setCorners(-1.0f, 1.0f, 1.0f, -1.0f)
rect.material = material
rect.renderQueueGroup = OgreJNI.RENDER_QUEUE_BACKGROUND_get().toShort()
val bb = AxisAlignedBox().apply { setInfinite() }
rect.boundingBox = bb

val node = scnMgr.rootSceneNode.createChildSceneNode()
node.attachObject(rect)
```

0v3rt1r3d commented:
Hm. If I create texture from resources with `material.techniques[0].passes[0].createTextureUnitState("Dirt.jpg")` it still disappear after some some frames rendered

paroj commented:
so you are rendering the background using ogre now?

anyway in the "PlaneMat" material will not work because it still needs a "samplerExternalOES" fragment shader.

The default generated shader is currently wrong as the RTSS does not now about that kind of samplers:
https://github.com/OGRECave/ogre/blob/1074eaab13d3856774f08aed50064b7b62d18011/Components/RTShaderSystem/src/OgreShaderGLSLProgramWriter.cpp#L56-L62

if you provide the right shader, ogre will not be able to correctly parse it:
https://github.com/OGRECave/ogre/blob/c5cae017771fda9d5d1ee7d848e5fa84d3695fa5/RenderSystems/GLES2/src/GLSLES/src/OgreGLSLESProgramManager.cpp#L69-L71

so you will not be able to change the sampler slot from 0, but that should be fine as it only uses one texture

paroj commented:
> it still disappear after some some frames rendered

I might have an idea about this one. Try making this a member variable
```
val rect = Rectangle2D(true)
```
rect is the only reference here. It must be kept somewhere so it does not get garbage collected.

paroj commented:
using "Dirt.jpg" as a background should be a prerequisite before we continue with TEX_TYPE_EXTERNAL_OES and custom shaders

0v3rt1r3d commented:
Well, after saving rect as a global variable background image stopped to disappear. Now I try to figure out how to make background rendering using ogre

paroj commented:
it boils down to to [setting this shader](https://github.com/google-ar/arcore-android-sdk/blob/master/samples/hello_ar_java/app/src/main/assets/shaders/screenquad.frag) on "PlaneMat" that you have created.
1. https://ogrecave.github.io/ogre/api/latest/class_ogre_1_1_high_level_gpu_program_manager.html#a992de8146fdd698d1cd6774fb15a2001
2. https://ogrecave.github.io/ogre/api/latest/class_ogre_1_1_pass.html#a3d4978eca2f3387a7abec740ec760ad7

paroj commented:
you have to put it into the `GLSLES/` subfolder and use `"glsles"` language identifier.

Furthermore you also must set a vertex shader. Here you can use [the upstream one](https://github.com/google-ar/arcore-android-sdk/blob/master/samples/hello_ar_java/app/src/main/assets/shaders/screenquad.vert), but you have to rename the attributes to match ogre convetions as:
https://ogrecave.github.io/ogre/api/latest/_high-level-_programs.html#Accessing-OpenGL-states-in-GLSL

0v3rt1r3d commented:
Yeah!!! Now I can see changing background, but I looks like one pixel is stretched to full screen

paroj commented:
please share a screenshot and your vertex shader. It sounds like your uv coordinates are all zero.

0v3rt1r3d commented:
It looks this way https://yadi.sk/i/xkR3YU8KoFB-RA, but background color is always changes and depends on the light in front of device

0v3rt1r3d commented:
No, this shader didn't help :(

paroj commented:
look out for shader compilation errors/ warnings in the log. Also try "visual debugging" by writing

```
gl_FragColor = vec4(v_TexCoord, 1, 1);
```

in the FS

0v3rt1r3d commented:
No compile errors happened. And "visual debugging" makes the whole screen blue

paroj commented:
how about
```glsl
attribute vec4 vertex;
attribute vec2 uv0;

varying vec2 v_TexCoord;

void main() {
   gl_Position = vertex;
   v_TexCoord = vertex.xy;
}
```

the issue is that `v_TexCoord` stays at (0,0) which is strange as "Dirt.jpg" w/o shaders worked

0v3rt1r3d commented:
We are close to solution, it looks this way: https://yadi.sk/i/9J_7HJmNRt0LkA

paroj commented:
> Also I think to add shaders I used. Will you review it?

yes, the shaders should go into https://github.com/OGRECave/ogre/tree/master/Samples/Media/materials/programs/GLSLES

also you should add a test app similar to
https://github.com/OGRECave/ogre/tree/master/Samples/AndroidJNI

0v3rt1r3d commented on RenderSystems/GLES2/src/OgreGLES2RenderSystem.cpp:
```diff
@@ -844,6 +844,23 @@ namespace Ogre {
 
         GLenum target = mTextureTypes[unit];
 
+        if (target == GL_TEXTURE_EXTERNAL_OES)
```
I do not clearly understand what I should do :(

paroj replied:
at the linked line:
```cpp
if (texPtr->getTextureType() == GL_TEXTURE_EXTERNAL_OES || texPtr->getTextureType() == TEX_TYPE_2D_RECT)
    setTextureAddressingMode( TAM_CLAMP );
else ..
```
instead of the check here

0v3rt1r3d commented:
I have some challenges with android test app.
1. As I figured out you have template for AndroidManifest.xml and generate several different projects using it with cmake. And just copy all other sources, resources and compiled library to the target project. Why don't you just create one gradle-project with different activities and put here your compiled .so library?
2. Maybe I did a mistake, but after building ogre, I didn't find any source files in SampleBrowserNDK project

paroj commented:
1. I did not yet come to port the Android build to gradle. Still using ant here. If it makes your life easier, you can also add another activity to the AndroidJNI sample.
Also feel free to modernize the Android build, but that would be out of scope for this PR.
2. SampleBrowserNDK is a pure C++ project and does not contain any Java sources. The TEX_TYPE_EXTERNAL_OES sample would not be part of it, but rather standalone - just like the AndroidJNI sample.

0v3rt1r3d commented:
Got it. Now I created created sample for oes texture, checked local build and try to build it on CI. Do you think it is ok that TEX_TYPE_EXTERNAL_OES is defined in OgreTexture.h, which means it is common for all render system textures (Direct, GL, Gl 3+, etc) but in practice it is available only for GLES2? I was fixing `'TEX_TYPE_EXTERNAL_OES' not handled in switch` errors, I figured out, what it would be better not to consider this type in other render systems at all. Can I do it using preprocessor's directives, like `#if <> #endif`? And if not, should I treat it as empty swtich-case with a comment or throw an error?

paroj commented on Samples/AndroidTextureOES/MainActivity.java:
```diff
@@ -0,0 +1,206 @@
+                1080,
+                1920,
+                0,
+                PixelFormat.PF_R8G8B8A8
```
currently it makes no difference, but this should be PF_UNKNOWN

paroj commented on .vscode/settings.json:
```diff
@@ -0,0 +1,5 @@
+{
```
please remove this file

paroj commented on Components/RTShaderSystem/src/OgreShaderGLSLProgramWriter.cpp:
```diff
@@ -60,6 +60,7 @@ void GLSLProgramWriter::initializeStringMaps()
     mGpuConstTypeMap[GCT_SAMPLERCUBE] = "samplerCube";
     mGpuConstTypeMap[GCT_SAMPLER1DSHADOW] = "sampler1DShadow";
     mGpuConstTypeMap[GCT_SAMPLER2DSHADOW] = "sampler2DShadow";
+    mGpuConstTypeMap[GCT_SAMPLER_EXTERNAL_OES] = "samplerExternalOES";
```
now that we got this, you could add writing
```cpp
#extension GL_OES_EGL_image_external : require
```
after this block

https://github.com/OGRECave/ogre/blob/1074eaab13d3856774f08aed50064b7b62d18011/Components/RTShaderSystem/src/OgreShaderGLSLESProgramWriter.cpp#L202-L207

so the RTSS auto-generates the correct shader and remove the custom shaders again. But this change is optional.

Also you would need to add the sampler here:

https://github.com/OGRECave/ogre/blob/1074eaab13d3856774f08aed50064b7b62d18011/Components/RTShaderSystem/src/OgreShaderGLSLESProgramWriter.cpp#L343-L344

paroj replied:
tried it.. GLSLES drivers dont like the FFP_SampleTexture indirection. Must get rid of it first

paroj commented:
> Can I do it using preprocessor's directives, like #if <> #endif? And if not, should I treat it as empty swtich-case with a comment or throw an error?

#ifdef is not an option as the idea of OgreMain is that you can write code that will always compile, but only work on certain rendersystems. (just like GL where some features might not be there)

I agree that TEX_TYPE_EXTERNAL_OES is quite ugly when it comes to passing to to code that does not support it. The only alternative I can think of is TextureUsage::TU_EXTERNAL.
However this will get ugly for code that *does* support it because it does not map well to the GL spec.

So for now I would stick with TEX_TYPE_EXTERNAL_OES and throw on all other rendersystems.
(as in `OGRE_EXCEPT(Exception::ERR_RENDERINGAPI_ERROR, "GL_OES_EGL_image_external unsupported", __FUNCTION__)`

paroj commented on Components/RTShaderSystem/src/OgreShaderFFPTexturing.cpp:
```diff
@@ -881,6 +881,9 @@ void FFPTexturing::setTextureUnit(unsigned short index, TextureUnitState* textur
         break;
+    case TEX_TYPE_EXTERNAL_OES:
+        curParams.mTextureSamplerType = GCT_SAMPLER_EXTERNAL_OES;
+        curParams.mVSInTextureCoordinateType = GCT_FLOAT2;
```
there is break missing here

paroj commented on CMake/Templates/AndroidManifest.xml.in:
```diff
@@ -5,6 +5,7 @@
     android:versionName="@OGRE_VERSION@" 
 	android:installLocation="auto">
 
+    <uses-permission android:name="android.permission.CAMERA" />
```
better make this `@OGRE_ADDITIONAL_PERMISSION@` and..

paroj commented on Samples/AndroidTextureOES/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+    SET(NDKOUT "${CMAKE_BINARY_DIR}/${ANDROID_MOD_NAME}")
+
+    SET(PKG_NAME "org.ogre.jni")
+
```
.. `set(OGRE_ADDITIONAL_PERMISSION "<uses-permission android:name="android.permission.CAMERA" />")` here.
So only this sample requests CAMERA permissions.

Otherwise our package here would require it too: https://f-droid.org/en/packages/org.ogre.browser/

paroj commented:
feel free to remove the WIP from the title if you are finished.

0v3rt1r3d commented:
Yes, will it enough to merge pr?

paroj commented:
:+1:
